Upgrade to TypeScript 4.7

Sadly `Or` and `Tuple` kinds are broken after upgrade, which will require some investigation whether this is usage error or a regression in TypeScript:
```
[12:57:02] failed: There were 42 TypeScript errors:
src/lib/core/property_mixins.ts:55:23 - error TS2322: Type 'Or<["solid" | "dashed" | "dotted" | "dotdash" | "dashdot", unknown]>' is not assignable to type 'PropertyAlias | PropertyConstructor<number[] | LineDash> | Kind<number[] | LineDash>'.
  Type 'Or<["solid" | "dashed" | "dotted" | "dotdash" | "dashdot", unknown]>' is not assignable to type 'Kind<number[] | LineDash>'.
    Types of property '__type__' are incompatible.
      Type 'unknown' is not assignable to type 'number[] | LineDash'.

55   line_dash:        [ k.Or(LineDash, k.Array(k.Number)), [] ],
(...)
```

--- a/bokehjs/make/package.json
+++ b/bokehjs/make/package.json
@@ -22,7 +22,7 @@
     "semver": "^7.3.7",
     "ts-node": "^10.9.1",
     "tsconfig-paths": "^4.1.0",
-    "typescript": "~4.6.2",
+    "typescript": "~4.8.3",
     "which": "^2.0.2",
     "yargs": "^17.4.0"
   }

--- a/bokehjs/make/tasks/defaults.ts
+++ b/bokehjs/make/tasks/defaults.ts
@@ -16,7 +16,7 @@ task("defaults:generate", () => {
   proc.stderr.on("data", (data) => {
     log(`${data}`)
   })
-  return new Promise((resolve, reject) => {
+  return new Promise<void>((resolve, reject) => {
     proc.on("error", reject)
     proc.on("exit", (code) => {
       if (code === 0)

--- a/bokehjs/src/compiler/compile.ts
+++ b/bokehjs/src/compiler/compile.ts
@@ -1,5 +1,5 @@
 import * as path from "path"
-import * as ts from "typescript"
+import ts from "typescript"
 import lesscss from "less"
 
 import {compiler_host, parse_tsconfig, default_transformers, compile_files, report_diagnostics, TSOutput, Inputs, Outputs} from "./compiler"

--- a/bokehjs/src/compiler/compiler.ts
+++ b/bokehjs/src/compiler/compiler.ts
@@ -1,5 +1,5 @@
 import chalk from "chalk"
-import * as ts from "typescript"
+import ts from "typescript"
 
 import {basename, dirname, join, relative} from "path"
 

--- a/bokehjs/src/compiler/linker.ts
+++ b/bokehjs/src/compiler/linker.ts
@@ -2,7 +2,7 @@ import {resolve, relative, join, dirname, basename, extname, normalize, sep} fro
 import module from "module"
 import crypto from "crypto"
 
-import * as ts from "typescript"
+import ts from "typescript"
 import * as terser from "terser"
 import chalk from "chalk"
 

--- a/bokehjs/src/compiler/package.json
+++ b/bokehjs/src/compiler/package.json
@@ -24,7 +24,7 @@
     "eslint": "^8.23.0",
     "less": "^4.1.3",
     "terser": "^5.15.0",
-    "typescript": "~4.6.2",
+    "typescript": "~4.8.3",
     "yargs": "^17.4.0"
   }
 }

--- a/bokehjs/src/compiler/transforms.ts
+++ b/bokehjs/src/compiler/transforms.ts
@@ -1,4 +1,4 @@
-import * as ts from "typescript"
+import ts from "typescript"
 
 export function apply<T extends ts.Node>(node: T, ...transforms: ts.TransformerFactory<T>[]): T {
   const result = ts.transform(node, transforms)
@@ -13,11 +13,12 @@ function is_require(node: ts.Node): node is ts.CallExpression {
 }
 
 export function relativize_modules(relativize: (file: string, module_path: string) => string | null) {
-  function relativize_specifier(source: ts.SourceFile, expr: ts.Expression | undefined): ts.StringLiteral | null {
+  function relativize_specifier(context: ts.TransformationContext, source: ts.SourceFile, expr: ts.Expression | undefined): ts.StringLiteral | null {
+    const {factory} = context
     if (expr != null && ts.isStringLiteralLike(expr) && expr.text.length > 0) {
       const relative = relativize(source.fileName, expr.text)
       if (relative != null)
-        return ts.createLiteral(relative)
+        return factory.createStringLiteral(relative)
     }
 
     return null
@@ -30,21 +31,21 @@ export function relativize_modules(relativize: (file: string, module_path: strin
 
         function visit(node: ts.Node): ts.Node {
           if (ts.isImportDeclaration(node)) {
-            const moduleSpecifier = relativize_specifier(root, node.moduleSpecifier)
+            const moduleSpecifier = relativize_specifier(context, root, node.moduleSpecifier)
             if (moduleSpecifier != null) {
-              const {decorators, modifiers, importClause, assertClause} = node
-              return factory.updateImportDeclaration(node, decorators, modifiers, importClause, moduleSpecifier, assertClause)
+              const {modifiers, importClause, assertClause} = node
+              return factory.updateImportDeclaration(node, modifiers, importClause, moduleSpecifier, assertClause)
             }
           }
           if (ts.isExportDeclaration(node)) {
-            const moduleSpecifier = relativize_specifier(root, node.moduleSpecifier)
+            const moduleSpecifier = relativize_specifier(context, root, node.moduleSpecifier)
             if (moduleSpecifier != null) {
-              const {decorators, modifiers, isTypeOnly, exportClause, assertClause} = node
-              return factory.updateExportDeclaration(node, decorators, modifiers, isTypeOnly, exportClause, moduleSpecifier, assertClause)
+              const {modifiers, isTypeOnly, exportClause, assertClause} = node
+              return factory.updateExportDeclaration(node, modifiers, isTypeOnly, exportClause, moduleSpecifier, assertClause)
             }
           }
           if (is_require(node)) {
-            const moduleSpecifier = relativize_specifier(root, node.arguments[0])
+            const moduleSpecifier = relativize_specifier(context, root, node.arguments[0])
             if (moduleSpecifier != null) {
               const {expression, typeArguments} = node
               return factory.updateCallExpression(node, expression, typeArguments, [moduleSpecifier])
@@ -80,7 +81,6 @@ export function insert_class_name() {
 
       if (ts.isClassDeclaration(node) && node.name != null && !has__name__(node)) {
         const property = factory.createPropertyDeclaration(
-          undefined,
           factory.createModifiersFromModifierFlags(ts.ModifierFlags.Static),
           "__name__",
           undefined,
@@ -89,7 +89,6 @@ export function insert_class_name() {
 
         node = factory.updateClassDeclaration(
           node,
-          node.decorators,
           node.modifiers,
           node.name,
           node.typeParameters,
@@ -105,7 +104,9 @@ export function insert_class_name() {
 }
 
 export function remove_use_strict() {
-  return (_context: ts.TransformationContext) => (root: ts.SourceFile) => {
+  return (context: ts.TransformationContext) => (root: ts.SourceFile) => {
+    const {factory} = context
+
     const statements = root.statements.filter((node) => {
       if (ts.isExpressionStatement(node)) {
         const expr = node.expression
@@ -115,7 +116,7 @@ export function remove_use_strict() {
       return true
     })
 
-    return ts.updateSourceFileNode(root, statements)
+    return factory.updateSourceFile(root, statements)
   }
 }
 
@@ -307,12 +308,14 @@ export function fix_esmodule() {
       return node
     })
 
-    return ts.updateSourceFileNode(root, statements)
+    return factory.updateSourceFile(root, statements)
   }
 }
 
 export function remove_void0() {
-  return (_context: ts.TransformationContext) => (root: ts.SourceFile) => {
+  return (context: ts.TransformationContext) => (root: ts.SourceFile) => {
+    const {factory} = context
+
     let found = false
     const statements = root.statements.filter((node) => {
       if (!found && ts.isExpressionStatement(node)) {
@@ -331,7 +334,7 @@ export function remove_void0() {
       return true
     })
 
-    return ts.updateSourceFileNode(root, statements)
+    return factory.updateSourceFile(root, statements)
   }
 }
 
@@ -373,19 +376,19 @@ export function fix_esexports() {
       return node
     })
 
-    return ts.updateSourceFileNode(root, statements)
+    return factory.updateSourceFile(root, statements)
   }
 }
 
 export function wrap_in_function(module_name: string) {
   return (context: ts.TransformationContext) => (root: ts.SourceFile) => {
     const {factory} = context
-    const p = (name: string) => factory.createParameterDeclaration(undefined, undefined, undefined, name)
+    const p = (name: string) => factory.createParameterDeclaration(undefined, undefined, name)
     const params = [p("require"), p("module"), p("exports"), p("__esModule"), p("__esExport")]
     const block = factory.createBlock(root.statements, true)
-    const func = factory.createFunctionDeclaration(undefined, undefined, undefined, "_", undefined, params, undefined, block)
+    const func = factory.createFunctionDeclaration(undefined, undefined, "_", undefined, params, undefined, block)
     ts.addSyntheticLeadingComment(func, ts.SyntaxKind.MultiLineCommentTrivia, ` ${module_name} `, false)
-    return ts.updateSourceFileNode(root, [func])
+    return factory.updateSourceFile(root, [func])
   }
 }
 

--- a/bokehjs/src/lib/core/build_views.ts
+++ b/bokehjs/src/lib/core/build_views.ts
@@ -8,7 +8,7 @@ export type Options<T extends View> = {parent: T["parent"] | null, owner?: ViewM
 
 async function _build_view<T extends HasProps>(view_cls: T["default_view"], model: T, options: Options<ViewOf<T>>): Promise<ViewOf<T>> {
   assert(view_cls != null, "model doesn't implement a view")
-  const view = new view_cls!({...options, model})
+  const view = new view_cls({...options, model})
   view.initialize()
   await view.lazy_initialize()
   return view

--- a/bokehjs/src/lib/core/kinds.ts
+++ b/bokehjs/src/lib/core/kinds.ts
@@ -109,9 +109,10 @@ export namespace Kinds {
     }
   }
 
-  export type TupleKind<T extends unknown[]> = {[K in keyof T]: T[K] extends T[number] ? Kind<T[K]> : never}
+  // See https://github.com/microsoft/TypeScript/issues/49556.
+  export type TupleKind<T extends unknown[]> = {[K in keyof T]: Kind<T[K]>}
 
-  export class Or<T extends unknown[]> extends Kind<T[number]> {
+  export class Or<T extends [unknown, ...unknown[]]> extends Kind<T[number]> {
     constructor(readonly types: TupleKind<T>) {
       super()
       this.types = types
@@ -447,7 +448,7 @@ export const Regex = (regex: RegExp) => new Kinds.Regex(regex)
 export const Null = new Kinds.Null()
 export const Nullable = <BaseType>(base_type: Kind<BaseType>) => new Kinds.Nullable(base_type)
 export const Opt = <BaseType>(base_type: Kind<BaseType>) => new Kinds.Opt(base_type)
-export const Or = <T extends unknown[]>(...types: Kinds.TupleKind<T>) => new Kinds.Or(types)
+export const Or = <T extends [unknown, ...unknown[]]>(...types: Kinds.TupleKind<T>) => new Kinds.Or(types)
 export const Tuple = <T extends [unknown, ...unknown[]]>(...types: Kinds.TupleKind<T>) => new Kinds.Tuple(types)
 export const Struct = <T extends object>(struct_type: {[key in keyof T]: Kind<T[key]>}) => new Kinds.Struct(struct_type)
 export const Arrayable = <ItemType>(item_type: Kind<ItemType>) => new Kinds.Arrayable(item_type)

--- a/bokehjs/src/lib/core/properties.ts
+++ b/bokehjs/src/lib/core/properties.ts
@@ -583,7 +583,7 @@ export class ColorSpec extends DataSpec<types.Color | null> {
     return encode_rgba(color2rgba(color))
   }
 
-  override v_materialize(colors: Arrayable<types.Color | null>): ColorArray {
+  override v_materialize(colors: Arrayable<types.Color | null> | NDArray): ColorArray {
     if (is_NDArray(colors)) {
       if (colors.dtype == "uint32" && colors.dimension == 1) {
         return to_big_endian(colors)

--- a/bokehjs/src/lib/document/defs.ts
+++ b/bokehjs/src/lib/document/defs.ts
@@ -21,7 +21,7 @@ export type KindRef =
   PrimitiveKindRef |
   ["Regex", string, string?] |
   ["Nullable", KindRef] |
-  ["Or", ...KindRef[]] |
+  ["Or", KindRef, ...KindRef[]] |
   ["Tuple", KindRef, ...KindRef[]] |
   ["Array", KindRef] |
   ["Struct", ...([string, KindRef][])] |
@@ -67,8 +67,8 @@ export function decode_def(def: ModelDef, deserializer: Deserializer): typeof Ha
           return kinds.Nullable(kind_of(subref))
         }
         case "Or": {
-          const [, ...subrefs] = ref
-          return kinds.Or(...subrefs.map(kind_of))
+          const [, subref, ...subrefs] = ref
+          return kinds.Or(kind_of(subref), ...subrefs.map(kind_of))
         }
         case "Tuple": {
           const [, subref, ...subrefs] = ref

